File: assignment2.py
def spiral_matrix(matrix):
    result = []
    top, bottom = 0, len(matrix)
    left, right = 0, len(matrix[0])
    while top < bottom and left < right:
        for i in range(left, right):
            result.append(matrix[top][i])
        top += 1
        for i in range(top, bottom):
            result.append(matrix[i][right-1])
        right -= 1
        if top < bottom:
            for i in range(right-1, left-1, -1):
                result.append(matrix[bottom-1][i])
            bottom -= 1
        if left < right:
            for i in range(bottom-1, top-1, -1):
                result.append(matrix[i][left])
            left += 1
    return result


matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
print("Spiral:", spiral_matrix(matrix))





def rotate90(matrix):
    return list(zip(*matrix[::-1]))


matrix = [[1,2,3],[4,5,6],[7,8,9]]
rotated = rotate90(matrix)
for row in rotated:
    print(list(row))
    
    
    
def sum_diagonals(matrix):
    n = len(matrix)
    main_diag = sum(matrix[i][i] for i in range(n))
    sec_diag = sum(matrix[i][n-i-1] for i in range(n))
    return main_diag + sec_diag - (matrix[n//2][n//2] if n % 2 else 0)

matrix = [[1,2,3],[4,5,6],[7,8,9]]
print("Sum of diagonals:", sum_diagonals(matrix))





matrix = [[1,2,3],[4,5,6],[7,8,9]]
transpose = [list(row) for row in zip(*matrix)]
for row in transpose:
    print(row)
    
    
    
def is_sparse(matrix):
    total = len(matrix) * len(matrix[0])
    zero_count = sum(row.count(0) for row in matrix)
    return zero_count > total / 2

matrix = [[0, 0, 3], [0, 0, 0], [0, 2, 0]]
print("Sparse Matrix" if is_sparse(matrix) else "Not Sparse")




a = [3, 2, 0, 5]  
b = [1, 0, 4]     
n = max(len(a), len(b))
res = [0]*n
for i in range(n):
    val1 = a[i] if i < len(a) else 0
    val2 = b[i] if i < len(b) else 0
    res[i] = val1 + val2
print("Sum:", res)





a = [1, 2, 3] 
b = [4, 5]     
res = [0] * (len(a) + len(b) - 1)
for i in range(len(a)):
    for j in range(len(b)):
        res[i+j] += a[i] * b[j]
print("Result:", res)



def is_identity(matrix):
    for i in range(len(matrix)):
        for j in range(len(matrix)):
            if (i == j and matrix[i][j] != 1) or (i != j and matrix[i][j] != 0):
                return False
    return True

matrix = [[1,0,0],[0,1,0],[0,0,1]]
print("Identity Matrix" if is_identity(matrix) else "Not Identity")







matrix = [[0, 0, 1], [0, 1, 2], [1, 2, 3]]
zero_count = sum(row.count(0) for row in matrix)
print("Zero count:", zero_count)

File: assignment2.c
#include <stdio.h>
void spiral(int a[10][10], int row, int col) {
    int top = 0, bottom = row, left = 0, right = col;
    while(top < bottom && left < right) {
        for(int i = left; i < right; i++) printf("%d ", a[top][i]);
        top++;
        for(int i = top; i < bottom; i++) printf("%d ", a[i][right-1]);
        right--;
        if(top < bottom) {
            for(int i = right-1; i >= left; i--) printf("%d ", a[bottom-1][i#include <stdio.h>

int main() {
    int mat[100][100], r, c, count = 0;
    scanf("%d %d", &r, &c);
    for (int i = 0; i < r; i++)
        for (int j = 0; j < c; j++) {
            scanf("%d", &mat[i][j]);
            if (mat[i][j] == 0) count++;
        }
    printf("%d", count);
    return 0;
}
            bottom--;
        }
        if(left < right) {
            for(int i = bottom-1; i >= top; i--) printf("%d ", a[i][left]);
            left++;
        }
    }
}
int main() {
    int a[10][10], row = 3, col = 3;
    for(int i=0; i<row; i++)
        for(int j=0; j<col; j++)
            scanf("%d", &a[i][j]);
    spiral(a, row, col);
    return 0;
}




#include <stdio.h>

void rotate90(int mat[100][100], int n) {
    for (int i = 0; i < n; i++)
        for (int j = i; j < n; j++) {
            int temp = mat[i][j];
            mat[i][j] = mat[j][i];
            mat[j][i] = temp;
        }

    for (int i = 0; i < n; i++)
        for (int j = 0, k = n - 1; j < k; j++, k--) {
            int temp = mat[i][j];
            mat[i][j] = mat[i][k];
            mat[i][k] = temp;
        }
}

int main() {
    int n, mat[100][100];
    scanf("%d", &n);
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) scanf("%d", &mat[i][j]);
    rotate90(mat, n);
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) printf("%d ", mat[i][j]);
        printf("\n");
    }
    return 0;
}





#include <stdio.h>

int main() {
    int n, mat[100][100], sum = 0;
    scanf("%d", &n);
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) scanf("%d", &mat[i][j]);

    for (int i = 0; i < n; i++) {
        sum += mat[i][i];
        if (i != n - i - 1) sum += mat[i][n - i - 1];
    }
    printf("%d", sum);
    return 0;
}








#include <stdio.h>

int main() {
    int r, c, mat[100][100];
    scanf("%d %d", &r, &c);
    for (int i = 0; i < r; i++)
        for (int j = 0; j < c; j++) scanf("%d", &mat[i][j]);

    for (int j = 0; j < c; j++) {
        for (int i = 0; i < r; i++) printf("%d ", mat[i][j]);
        printf("\n");
    }
    return 0;
}







#include <stdio.h>

int main() {
    int r, c, mat[100][100], count = 0;
    scanf("%d %d", &r, &c);
    for (int i = 0; i < r; i++)
        for (int j = 0; j < c; j++) {
            scanf("%d", &mat[i][j]);
            if (mat[i][j] == 0) count++;
        }

    if (count > (r * c) / 2)
        printf("Sparse matrix");
    else
        printf("Not a sparse matrix");

    return 0;
}





#include <stdio.h>

int main() {
    int a[100], b[100], res[100], n;
    scanf("%d", &n); 
    for (int i = 0; i < n; i++) scanf("%d", &a[i]);
    for (int i = 0; i < n; i++) scanf("%d", &b[i]);
    for (int i = 0; i < n; i++) res[i] = a[i] + b[i];
    for (int i = 0; i < n; i++) printf("%d ", res[i]);
    return 0;
}





#include <stdio.h>

int main() {
    int a[100], b[100], res[200] = {0}, n1, n2;
    scanf("%d", &n1);
    for (int i = 0; i < n1; i++) scanf("%d", &a[i]);
    scanf("%d", &n2);
    for (int i = 0; i < n2; i++) scanf("%d", &b[i]);
    for (int i = 0; i < n1; i++)
        for (int j = 0; j < n2; j++)
            res[i + j] += a[i] * b[j];
    for (int i = 0; i < n1 + n2 - 1; i++) printf("%d ", res[i]);
    return 0;
}




#include <stdio.h>
#include <stdlib.h>

int main() {
    int *arr, n, i;
    scanf("%d", &n);
    arr = (int *)malloc(n * sizeof(int));
    for (i = 0; i < n; i++) scanf("%d", &arr[i]);

    
    for (i = n - 1; i >= 0; i--) printf("%d ", arr[i]);

    free(arr);
    return 0;
}



#include <stdio.h>

int main() {
    int n, mat[100][100], flag = 1;
    scanf("%d", &n);
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++) {
            scanf("%d", &mat[i][j]);
            if ((i == j && mat[i][j] != 1) || (i != j && mat[i][j] != 0))
                flag = 0;
        }

    if (flag)
        printf("Identity matrix");
    else
        printf("Not an identity matrix");

    return 0;
}




#include <stdio.h>

int main() {
    int mat[100][100], r, c, count = 0;
    scanf("%d %d", &r, &c);
    for (int i = 0; i < r; i++)
        for (int j = 0; j < c; j++) {
            scanf("%d", &mat[i][j]);
            if (mat[i][j] == 0) count++;
        }
    printf("%d", count);
    return 0;
}




